Overhaul sequence filename definitions

pchote opened the pull request:
This is the first in a series of PRs aiming to remove technical debt from the sequence code and add features needed to finish and upstream remastered asset support.

All the magic around sprite filenames is removed in favour of explicitly specifying the full sprite name (per tileset if needed). This allows mods that don't follow the existing C&C conventions (e.g. by putting each tilesets assets in its own directory) to use tileset-specific artwork. It also prepares for a future PR that will add new ways of specifying filenames (e.g. `FilenamePattern: mytank-{:D4}.png`).

The `Defaults` node was previously combined into each sequence using `MiniYaml.Merge` at sequence-parsing time. This prevented us from implementing recursive merging at the `MiniYaml` level (see e.g. [this complaint](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/GameRules/WeaponInfo.cs#L141)) because the raw sequence yaml may be malformed (e.g. trying to remove keys that don't exist) before the defaults have been copied over. This has been replaced with a much simpler fallback convention: the parser first looks for the requested key in the current sequence, and if not found it will look again in `Defaults`. Sequences can override (or more correctly, mask) the `Defaults` value by defining a key with no value.

I suggest reviewing by commit, and reading the commit descriptions for more information.
The yaml changes in the third commit are fully automated, so it shouldn't be necessary to inspect every single line in detail. A small number of manual cleanups are specified in the fourth commit.

pchote committed:
Reimplement sequence Defaults parsing:


The previous MiniYaml.Merge implementation interacted
poorly with yaml inheritance, making it complicated
(or impossible) to override certain keys from Defaults.

The new implementation is simpler: If a key is defined
it will be used. If it isn't, the default (if defined)
will be used. Defaults can be masked by making sure
the same key is defined (even with an empty value)
in the sequence.

This also fixes naming within the sequence code to
distinguish between images (a group of sequences),
sequences (defining a specific sprite/animation),
and filenames for a specific sprite/animation.

PunkPun commented on OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:
```diff
@@ -196,9 +228,19 @@ public static List<string> UpdateMod(ModData modData, UpdateRule rule, out YamlF
 			manualSteps.AddRange(ApplyTopLevelTransform(modData, modRules, rule.UpdateActorNode));
+
+			var resolvedWeapons = MiniYaml.Load(modData.DefaultFileSystem, modData.Manifest.Weapons, null);
+			manualSteps.AddRange(rule.BeforeUpdateSequences(modData, resolvedWeapons));
```
```suggestion
			manualSteps.AddRange(rule.BeforeUpdateWeapons(modData, resolvedWeapons));
```

PunkPun commented on OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:
```diff
@@ -114,6 +114,9 @@ public static List<string> UpdateMap(ModData modData, IReadWritePackage mapPacka
 				if (mapRulesNode != null)
 				{
+					var resolvedActors = LoadMapYaml(modData.DefaultFileSystem, mapPackage, modData.Manifest.Rules, mapRulesNode.Value);
+					manualSteps.AddRange(rule.BeforeUpdateWeapons(modData, resolvedActors));
```
```suggestion
					manualSteps.AddRange(rule.BeforeUpdateActors(modData, resolvedActors));
```

PunkPun reviewed (commented):
Since we are changing everything up we may want a more user friendly exception for missing values. It doesn't mention where something is missing nor what

I deleted the filename field from MiG and got this
```
Exception of type `System.IO.FileNotFoundException`: File not found: 
   at OpenRA.Mods.Common.Graphics.FileNotFoundSequence.OpenRA.Graphics.ISpriteSequence.get_Tick() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:line 112
   at OpenRA.Graphics.Animation.PlaySequence(String sequenceName) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Graphics/Animation.cs:line 126
   ```
   
   Completely unhelpful to modders

PunkPun commented:
The CheckSequences lint test (or a new one) could also be used to pick up on missing values

pchote commented:
Fixed.

PunkPun commented:
This is much better 👍. However it would be more ideal if lint test didn't display the tracelogs

pchote commented:
The next PR completely overhauls the way that the sprite loading and lint tests work, so can we leave that for now?

PunkPun commented:
sure

PunkPun commented:
I tried updating [Treasure_Hunt.oramap.zip](https://github.com/OpenRA/OpenRA/files/10300023/Treasure_Hunt_BI-3.5-2.oramap.zip) and the update rule crashed

<details>
  <summary>Crashlog</summary>

```
System.Collections.Generic.KeyNotFoundException: The given key 'SNOW' was not present in the dictionary.
        at OpenRA.Mods.Common.UpdateRules.Rules.ExplicitSequenceFilenames.ProcessNode(ModData modData, MiniYamlNode sequenceNode, MiniYamlNode resolvedSequenceNode, String imageName) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:line 281
        at OpenRA.Mods.Common.UpdateRules.Rules.ExplicitSequenceFilenames.UpdateSequenceNode(ModData modData, MiniYamlNode imageNode)+MoveNext() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:line 136
        at OpenRA.Mods.Common.UpdateRules.UpdateUtils.ApplyTopLevelTransform(ModData modData, List`1 files, TopLevelNodeTransform transform)+MoveNext() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:line 292
        at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
        at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
        at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMap(ModData modData, IReadWritePackage mapPackage, UpdateRule rule, List`1& files, HashSet`1 externalFilenames) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UpdateRules/UpdateUtils.cs:line 143
        at OpenRA.Mods.Common.UtilityCommands.UpdateMapCommand.ApplyRules(ModData modData, IReadWritePackage mapPackage, IEnumerable`1 rules) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/UtilityCommands/UpdateMapCommand.cs:line 108
```
</details>

pchote commented:
The update rule requires the legacy tileset metadata (which is removed in the last commit) to be defined in mod.yaml. The map updates fine if you add that back.

PunkPun commented:
Seems less than ideal. Update rules are also meant for maps, I'd even say that they're used for maps more than for mods. We should handle the crashes

pchote commented:
I can’t think of an acceptable way to do that - keeping unused junk data in mod.yaml isn’t acceptable, and hardcoding the upgrade rule with data for only the upstream mods isn’t really either.

PunkPun commented on OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:
```diff
@@ -0,0 +1,378 @@
+						                   filename.Substring(2, filename.Length - 2);
+
+					if (addExtension)
+						overrideFilename += useTilesetExtension ? tilesetExtensions[sequenceTileset] : defaultSpriteExtension;
```
We could just check if the `sequenceTileset` exists and if it doesn't just not do the step. Instead we could log a note for the modder to fix it manually

PunkPun replied:
The same issue could happen with TilesetCodes as well correct?

abcdefg30 commented:
The update rule shouldn't crash but write a proper error message - explaining what is written in the comments above: That it needs the definitions in mod.yaml and that those can be removed (again) afterwards.

pchote commented:
Updated.

```
ExplicitSequenceFilenames: Sequence filenames must be specified explicitly.
   Updating map... COMPLETE
   Manual changes are required to complete this update:
    * The ExplicitSequenceFilenames rule requires TilesetExtensions
      to be defined under the SpriteSequenceFormat definition in mod.yaml.
      Add these definitions back and run the update rule again.

Semi-automated update complete.
Please review the messages above for any manual actions that must be applied.
```

pchote committed:
Manual yaml cleanup.

PunkPun commented:
In the future we should make it clearer when update rules fail / require manual changes. Currently that information is fairly hard to find within walls of text.

I did run this rule on some other fairly artwork heavy maps and it seems to function well

PunkPun reviewed (dismissed):
LGTM

PunkPun commented:
Could there be anything done to the resource centre update function to account for the missing definitions?

pchote commented:
Yes, the definitions can be added back to mod.yaml on the RC engines.

penev92 commented on OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs:
```diff
@@ -20,9 +20,9 @@ public class ClassicSpriteSequenceLoader : DefaultSpriteSequenceLoader
 			: base(modData) { }
 
-		public override ISpriteSequence CreateSequence(ModData modData, string tileSet, SpriteCache cache, string sequence, string animation, MiniYaml info)
+		public override ISpriteSequence CreateSequence(ModData modData, string tileset, SpriteCache cache, string image, string sequence, MiniYaml data, MiniYaml defaults)
```
I am not convinced `sequence` -> `image` and `animation` -> `sequence` makes it better. My recommendation for "SequenceCollection" + "Sequence" (or "Animation"?) still stands.
To me *"image"* was the problem we had before, not the solution.

pchote replied:
`Sequence` and `Image` are the names used by traits. This PR is already big and rebase-prone, I *really* don't want to have to edit another 50+ trait files!

PunkPun commented:
Needs a rebase / copyright header update

pchote commented:
Rebased.

Mailaender commented on mods/ts/sequences/vehicles.yaml:
```diff
@@ -1,189 +1,242 @@
 	Defaults:
 		Offset: 0, 0, 24
 	muzzle:
+		Filename:
```
Why is this required?

Mailaender commented on OpenRA.Mods.Cnc/Graphics/ClassicSpriteSequence.cs:
```diff
@@ -33,15 +33,13 @@ public class ClassicSpriteSequence : DefaultSpriteSequence
 
 			if (useClassicFacings && facings != 32)
-				throw new InvalidOperationException(
-					$"{info.Nodes[0].Location}: Sequence {sequence}.{animation}: UseClassicFacings is only valid for 32 facings");
```
Why are you removing the line number?

pchote replied:
`info.Nodes[0].Location` was often the wrong line! (and the current plumbing doesn't let us easily report the correct one)

This PR removes it for correctness and consistency with the other sequence yaml errors.
#20597 implements line numbers properly for all errors.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -41,49 +41,49 @@ public bool TryGetPaletteForFrame(int frame, out uint[] palette)
+			foreach (var sequenceNode in imageNode.Value.Nodes)
 			{
-				using (new Support.PerfTimer($"new Sequence(\"{node.Key}\")", 20))
+				using (new Support.PerfTimer($"new Sequence(\"{imageNode.Key}\")", 20))
```
Unrelated, but isn't this supposed to be *outside* of the `foreach`?
Also unrelated, but the `return new ReadOnlyDictionary<string, ISpriteSequence>(sequences);` at the end is redundant. Should just be `return sequences;` and the method's signature will ensure consumers only see an `IReadOnlyDictionary<>`.

pchote replied:
These will be dealt with in #20597.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -414,7 +417,10 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
-					var subFilename = GetSpriteFilename(modData, tileSet, combineSequenceNode.Key, sequence, combineData, NoData);
+					var subFilename = GetSpriteFilename(modData, tileSet, image, sequence, combineData, NoData);
+					if (subFilename == null)
+						throw new YamlException($"Sequence {image}.{sequence}.{combineSequenceNode.Key} does not define a filename");
```
Full stop at the end of the sentence please.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -440,6 +446,8 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 				// Different sequences may apply different offsets to the same frame
 				var filename = GetSpriteFilename(modData, tileSet, image, sequence, data, defaults);
+				if (filename == null)
+					throw new YamlException($"Sequence {image}.{sequence} does not define a filename");
```
Same + empty line below.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -497,6 +505,9 @@ public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache,
 			{
 				var filename = GetSpriteFilename(modData, tileSet, image, sequence, data, defaults);
+				if (filename == null)
+					throw new YamlException($"Sequence {image}.{sequence} does not define a filename");
```
Full stop at the end of the sentence please.

penev92 commented on OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:
```diff
@@ -0,0 +1,412 @@
+		public override string Name => "Sequence filenames must be specified explicitly.";
+
+		public override string Description =>
+			"Sequence sprite filenames are no longer automatically inferred, and the AddExtension.\n" +
```
```suggestion
			"Sequence sprite filenames are no longer automatically inferred and the AddExtension,\n" +
```

penev92 commented on OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:
```diff
@@ -0,0 +1,412 @@
+
+		public override string Description =>
+			"Sequence sprite filenames are no longer automatically inferred, and the AddExtension.\n" +
+			"UseTilesetExtension, UseTilesetNodes, and TilesetOverrides fields have been removed.\n\n" +
```
```suggestion
			"UseTilesetExtension, UseTilesetNodes and TilesetOverrides fields have been removed.\n\n" +
```

penev92 commented on OpenRA.Mods.Common/UpdateRules/Rules/20221203/ExplicitSequenceFilenames.cs:
```diff
@@ -0,0 +1,412 @@
+				if (!string.IsNullOrEmpty(sequenceNode.Value.Value))
+					explicitlyNamedSequences.Add(sequenceNode.Key);
+
+				var resolvedSequenceNode = resolvedImageNode.Value.Nodes.SingleOrDefault(n => n.Key == sequenceNode.Key);
```
This will return `null` for node removals, causing an NRE in `ProcessNode` below. Please add a check above to ignore if this is a removal, like you do for "Inherits".

penev92 reviewed (changes requested):
The naming may or may not just be me, but there is at least one crash to fix.

PunkPun commented:
https://github.com/OpenRA/OpenRA/wiki/Sequences
will need to be updated after the merge

pchote committed:
Simplify tileset-specific sequence definitions.

All magic behaviour for constructing sprite filenames
has been removed in favour of an explicit Filename
(and TilesetFilenames for tileset-specific sequences)
property.

pchote commented:
Updated.

Mailaender reviewed (dismissed):
`--check-sprite-sequences` clean. No problems in-game.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -41,49 +41,49 @@ public bool TryGetPaletteForFrame(int frame, out uint[] palette)
 
 	public class DefaultSpriteSequenceLoader : ISpriteSequenceLoader
 	{
+		static readonly MiniYaml NoData = new MiniYaml(null);
```
This breaks `ExtractSpriteSequenceDocsCommand` ☹️ because apparently if you call `GetGenericTypeDefinition()` on a non-generic type it throws.
Please change
```diff
Properties = type.GetFields(BindingFlags.NonPublic | BindingFlags.Static)
-	.Where(fi => fi.FieldType.GetGenericTypeDefinition() == typeof(SpriteSequenceField<>))
+	.Where(fi => fi.FieldType.IsGenericType && fi.FieldType.GetGenericTypeDefinition() == typeof(SpriteSequenceField<>))
	.Select(fi =>
```
in `ExtractSpriteSequenceDocsCommand`, line 65.

pchote committed:
Fix ExtractSpriteSequenceDocsCommand crash on non-generic static fields.

pchote commented:
Updated.

penev92 reviewed (approved):
👍

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1cd13a6e3b0d3f258483ba154af6962856106cc4...0ebf015d53056e47d3c82b836ce58e3266a81c50)
